File: sdk-ms-azure-management-demo/RmsApiKeyCredentials.cs
using Microsoft.Rest;
using Newtonsoft.Json;
using System.Text;

namespace VodCreatorApp
{
    public class RmsApiKeyCredentials : ServiceClientCredentials
    {
        private readonly Uri _apiEndpoint;
        private readonly Uri _authorityUri;
        private readonly string? _subscriptionId;
        private readonly string? _apiKey;
        private DateTime tokenValidTo = DateTime.MinValue;

        private string _token = string.Empty;

        public RmsApiKeyCredentials(Uri apiEndpoint, string subscriptionId, string apiKey)
        {
            _apiEndpoint = apiEndpoint;
            var authorityUri = new Uri(_apiEndpoint, "/auth/token");
            _authorityUri = authorityUri;
            _subscriptionId = subscriptionId;
            _apiKey = apiKey;
        }

        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            if (DateTime.UtcNow > tokenValidTo)
            {
                var tokenRequest = new GetTokenRequest
                {
                    SubscriptionId = _subscriptionId,
                    ApiKey = _apiKey,
                };

                var authContent = new StringContent(JsonConvert.SerializeObject(tokenRequest), Encoding.UTF8, "application/json");
                var authResponse = await new HttpClient().PostAsync(_authorityUri, authContent, cancellationToken);

                authResponse.EnsureSuccessStatusCode();

                _token = await authResponse.Content.ReadAsStringAsync(cancellationToken);

                string tokenDataPart = _token[(_token.IndexOf('.') + 1)..];
                tokenDataPart = tokenDataPart[..tokenDataPart.IndexOf('.')];
                string tokenDataStr = Encoding.UTF8.GetString(Convert.FromBase64String(tokenDataPart.PadRight(4 * ((tokenDataPart.Length + 3) / 4), '=')));

                dynamic tokenData = JsonConvert.DeserializeObject<dynamic>(tokenDataStr);

                tokenValidTo = DateTime.UnixEpoch.AddSeconds((int)tokenData.exp);
            }

            request.Headers.Add("Authorization", $"Bearer {_token}");
            request.RequestUri = new Uri(_apiEndpoint, request.RequestUri.PathAndQuery);
            await base.ProcessHttpRequestAsync(request, cancellationToken);
        }
    }

    public class GetTokenRequest
    {
        public string? SubscriptionId { get; set; }

        public string? ApiKey { get; set; }
    }
}

File: node-js-demo/RmsApiKeyCredentials.js
const fetch = require('node-fetch');

class RmsApiKeyCredentials {
  constructor(authorityUri, subscriptionId, apiKey) {
    if (!authorityUri.endsWith('/auth/token')) {
      authorityUri += '/auth/token';
    }

    this.authorityUri = authorityUri;
    this.subscriptionId = subscriptionId;
    this.apiKey = apiKey;
  }

  async getToken(scopes, options) {
    const tokenRequest = {
      subscriptionId: this.subscriptionId,
      apiKey: this.apiKey,
    };

    const authResponse = await fetch(this.authorityUri, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(tokenRequest)
    });

    if (!authResponse.ok) {
      const responseText = await authResponse.text();
      throw new Error(
        `Failed to retrieve access token: ${authResponse.statusText}, ${responseText}`
      );
    }

    const tokenData = await authResponse.text();
    const tokenValidTo = new Date(0);
    tokenValidTo.setUTCSeconds(tokenData.exp);

    return {
      token: tokenData,
      expiresOnTimestamp: tokenValidTo.getTime()
    };
  }
}

module.exports = RmsApiKeyCredentials;
